ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -62,16 +81,62 @@ def get_waypoint_velocity(self, waypoint):
     def set_waypoint_velocity(self, waypoints, waypoint, velocity):
         waypoints[waypoint].twist.twist.linear.x = velocity
 
-    def distance(self, waypoints, wp1, wp2):
-        dist = 0
-        dl = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2  + (a.z-b.z)**2)
-        for i in range(wp1, wp2+1):
-            dist += dl(waypoints[wp1].pose.pose.position, waypoints[i].pose.pose.position)
-            wp1 = i
-        return dist
+    def distance_to_previous(self, position):
+        return euclidean(get_position(self.previous_pos), get_position(position))
+
+    def waypoint_is_ahead(self, waypoint):
+        w_x, w_y, _ = get_position(waypoint.pose.pose)
+        p_x, p_y, _ = get_position(self.previous_pos)
+        heading = math.atan2(w_y-p_y, w_x-p_x)
+        yaw = euler_from_quaternion(get_orientation(self.previous_pos))[2]
+        angle = normalize_angle(yaw-heading)
+        return True if math.fabs(angle) < math.pi/4. else False
+
+    def update_next_waypoint(self):
+        it = 0
+        nb_waypoints = len(self.static_waypoints.waypoints)
+        while not self.waypoint_is_ahead(self.static_waypoints.waypoints[self.next_waypoint % nb_waypoints]) and \
+                it < UPDATE_MAX_ITER:
+            it += 1
+            self.next_waypoint += 5  # We look at the next one
```

Comment:
Should be `+1` ?

Reply:
Extracted it to a separate issue https://github.com/clockworkbits/CarND-Capstone/issues/12